scripts/west_commands/runners/nrfutil.py

```diff
@@ -0,0 +1,122 @@
+                 tool_opt=[], force=False, recover=False):
+
+        super().__init__(cfg, family, softreset, dev_id, erase,
+                         tool_opt, force, recover)
+        self._ops = []
+        self._op_id = 1
+
+    @classmethod
+    def name(cls):
+        return 'nrfutil'
+
+    @classmethod
+    def tool_opt_help(cls) -> str:
+        return 'Additional options for nrfutil, e.g. "--log-level"'
+
+    @classmethod
+    def do_create(cls, cfg, args):
+        return NrfUtilBinaryRunner(cfg, args.nrf_family, args.softreset,
+                                   args.dev_id, erase=args.erase,
+                                   tool_opt=args.tool_opt, force=args.force,
+                                   recover=args.recover)
+
+    def _exec(self, args):
+        try:
+            out = self.check_output(['nrfutil', '--json', 'device'] + args)
+        except subprocess.CalledProcessError as e:
+            # https://docs.python.org/3/reference/compound_stmts.html#except-clause
+            cpe = e
+            out = cpe.stdout
+        else:
+            cpe = None
+        finally:
+            # https://github.com/ndjson/ndjson-spec
+            out = [json.loads(s) for s in
+                   out.decode(sys.getdefaultencoding()).split('\n') if len(s)]
+            self.logger.debug(f'output: {out}')
+            if cpe:
+                if 'execute-batch' in args:
+                    for o in out:
+                        if o['type'] == 'batch_end' and o['data']['error']:
+                            cpe.returncode = o['data']['error']['code']
+                raise cpe
+
+        return out
+
+    def do_get_boards(self):
+        out = self._exec(['list'])
+        devs = []
+        for o in out:
+            if o['type'] == 'task_end':
+                devs = o['data']['data']['devices']
+        snrs = [dev['serialNumber'] for dev in devs]
+
+        self.logger.debug(f'Found boards: {snrs}')
+        return snrs
+
+    def do_require(self):
+        self.require('nrfutil')
+
+    def _insert_op(self, op):
+        op['operationId'] = f'{self._op_id}'
+        self._op_id += 1
+        self._ops.append(op)
+
+    def _exec_batch(self):
+        # prepare the dictionary and convert to JSON
+        batch = json.dumps({'family': f'{self.family}',
+                            'operations': [op for op in self._ops]},
+                            indent=4) + '\n'
+        fd, name = tempfile.mkstemp(prefix='nrfutil-batch-',
+                                    suffix='.json', text=True)
+        try:
+            os.write(fd, batch.encode('utf-8'))
+        finally:
+            os.close(fd)
+        # reset first in case an exception is thrown
+        self._ops = []
+        self._op_id = 1
+        self.logger.debug(f'Executing batch in: {name}')
+        out = self._exec(['execute-batch', '--batch-path', f'{name}',
+                          '--serial-number', f'{self.dev_id}'])
```

Comment:
is temp file automatically removed? Btw, `with tempfile.NamedTemporaryFile() ...` could be used as well.

Reply:
It is not automatically removed. The problem with `tempfile.NamedTemporaryFile()` is:

> Whether the name can be used to open the file a second time, while the named temporary file is still open, varies across platforms (it can be so used on Unix; it cannot on Windows)

I want the file to remain on the filesystem after the execution is complete, because all of the logic is there in the JSON file, so in order to find out what went wrong I want to be able to `cat` the tempfile after execution is complete.